PowerShellTools/Intellisense/BraceCompletionController.cs

```diff
@@ -0,0 +1,287 @@
+                        return VSConstants.S_OK;
+                    }
+                    SetBraceCompleteState(false);
+                    break;
+                case (uint)VSConstants.VSStd2KCmdID.DELETE:                    
+                case (uint)VSConstants.VSStd2KCmdID.UNDO:
+                    SetBraceCompleteState(false);
+                    break;
+                default:
+                    break;
+            }
+
+            return NextCommandHandler.Exec(ref pguidCmdGroup, nCmdID, nCmdexecopt, pvaIn, pvaOut);
+        }
+
+        public int QueryStatus(ref Guid pguidCmdGroup, uint cCmds, OLECMD[] prgCmds, IntPtr pCmdText)
+        {
+            return NextCommandHandler.QueryStatus(ref pguidCmdGroup, cCmds, prgCmds, pCmdText);
+        }
+
+        #endregion
+        
+        /// <summary>
+        /// Complete the left brace with matched right brace
+        /// </summary>
+        /// <param name="leftBrace">The left brace to be completed.</param>
+        private void CompleteBrace(char leftBrace)
+        {
+            var typedCharToString = leftBrace.ToString();
+
+            // Step 1, create an undo transaction in case after typing the left brace user then click BACKSPACE to delete it. 
+            //         In that case, we need to undo this transaction, which will delete not only the left brace but also the filled right brace
+            // Step 2, insert the typed char which is the left brace
+            // Step 3, insert a matched right brace and then move caret to the previous char, meaning the caret ends up in the middle of the brace pair.
+            using (var undo = _undoHistory.CreateTransaction("Fill in " + typedCharToString))
+            {
+                _editorOperations.AddBeforeTextBufferChangePrimitive();
+
+                _editorOperations.InsertText(typedCharToString);
+                _editorOperations.InsertText(GetMatchedRightBrace(leftBrace).ToString());
+                _editorOperations.MoveToPreviousCharacter(false);
+
+                _editorOperations.AddAfterTextBufferChangePrimitive();
+                undo.Complete();
+            }     
+        }
+        
+        private bool ProcessTypedRightBrace(char typedChar)
+        {
+            if (_isLastCmdBraceComplete)
+            {
+                _editorOperations.MoveToNextCharacter(false);
+            }
+
+            return _isLastCmdBraceComplete;
+        }
+
+        private bool ProcessReturnKey()
+        {
+            bool isReturnKeyProcessed = IsCaretInMiddleOfPairedCurlyBrace();
+            if (isReturnKeyProcessed)
+            {
+                using (var undo = _undoHistory.CreateTransaction("Insert new line."))
+                {
+                    _editorOperations.AddBeforeTextBufferChangePrimitive();
+
+                    DeleteRightBrace();
+                    _editorOperations.InsertNewLine();
+                    _editorOperations.InsertText('}'.ToString());

```

Comment:
The ToString is extraneous ☺  I don’t (think) the compiler will optimize this away.

Until next adventure,

-Andre

From: Huanhuan Sun [mailto:notifications@github.com]
Sent: Monday, February 9, 2015 11:36 AM
To: Microsoft/poshtools
Cc: Andre Sayre
Subject: Re: [poshtools] Brace match (#117)


In PowerShellTools/Intellisense/BraceCompletionController.cs<https://github.com/Microsoft/poshtools/pull/117#discussion_r24357648>:

> +

> +            return _isLastCmdBraceComplete;

> +        }

> +

> +        private bool ProcessReturnKey()

> +        {

> +            bool isReturnKeyProcessed = IsCaretInMiddleOfPairedCurlyBrace();

> +            if (isReturnKeyProcessed)

> +            {

> +                using (var undo = _undoHistory.CreateTransaction("Insert new line."))

> +                {

> +                    _editorOperations.AddBeforeTextBufferChangePrimitive();

> +

> +                    DeleteRightBrace();

> +                    _editorOperations.InsertNewLine();

> +                    _editorOperations.InsertText('}'.ToString());

I originally wanted keep them consistent or consider them as Constant. Though don't have a strong preference here.

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/poshtools/pull/117/files#r24357648>.